### ISSUE TITLE
Embedding values from this library not matching SentenceTransformers

### ISSUE THREAD

**Issue by dhruv-anand-aintech** [taylorai/mlx_embedding_models#3]
Hi, I tried comparing the values from here with those from SentenceTransformers and they are not the same:
```
from sentence_transformers import SentenceTransformer

model2 = SentenceTransformer("TaylorAI/bge-micro-v2")


from mlx_embedding_models.embedding import EmbeddingModel
model = EmbeddingModel.from_registry("bge-small")
texts = [
    "isn't it nice to be inside such a fancy computer",
    "the horse raced past the barn fell"
]
embs = model.encode(texts)
print(embs.shape)
embs2 = model2.encode(texts)
# check if the embeddings are the same
import numpy as np

print(np.allclose(embs, embs2, atol=1e-5))
```

Output is false. Increasing atol doesn't work.

**Comment by andersonbcdefg** [comment]
This seems to be specific to bge-micro (I get True with bge-small) which makes me think this might be a pooling or configuration bug specific to this model. Looking into it.

**Comment by andersonbcdefg** [comment]
Confirmed this is a pooling bug. This model uses mean-pooling, but when it's pulled from HuggingFace with SentenceTransformers, it somehow ends up with a class-pool instead, even though the configuration says it should use mean pooling in 1_Pooling/config.json.

**Comment by andersonbcdefg** [comment]
Fixing pooling didn't resolve this yet, gotta dig deepers. Sorry! In the meantime, confirmed bge-small works so use that if you can.

**Comment by dhruv-anand-aintech** [comment]
Thanks for looking into it! 

FYI, I have integrated this library into Vector-io to reembed text data with different models on Mac.

**Comment by dhruv-anand-aintech** [comment]
https://github.com/AI-Northstar-Tech/vector-io/blob/main/src%2Fvdf_io%2Fscripts%2Freembed.py#L548-L560

**Comment by dhruv-anand-aintech** [comment]
a quick rerun gives this error:
```
{
	"name": "AttributeError",
	"message": "module 'mlx.core.metal' has no attribute 'clear_cache'",
	"stack": "---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[1], line 12
      7 model = EmbeddingModel.from_registry(\"bge-micro\")
      8 texts = [
      9     \"isn't it nice to be inside such a fancy computer\",
     10     \"the horse raced past the barn fell\"
     11 ]
---> 12 embs = model.encode(texts)
     13 print(embs.shape)
     14 embs2 = model2.encode(texts)

File ~/miniforge3/lib/python3.10/site-packages/mlx_embedding_models/embedding.py:259, in EmbeddingModel.encode(self, sentences, batch_size, show_progress, **kwargs)
    257         del batch
    258     # we're done with this seqlen, clear the cache
--> 259     mx.metal.clear_cache()
    261 # concatenate embeddings and reverse the sort
    262 output_embeddings = mx.concatenate(output_embeddings, axis=0)

AttributeError: module 'mlx.core.metal' has no attribute 'clear_cache'"
}
```

debugging further...

**Comment by andersonbcdefg** [comment]
Did you upgrade mlx_embedding_models to the latest version (now 0.0.9)? Does it have significant error still? (interested in the allclose but also the cosine similarity

**Comment by dhruv-anand-aintech** [comment]
Yes, it's the latest

**Comment by jyc** [comment]
This library's outputs for the bge-large and bge-m3 models also don't match the corresponding outputs from the FlagEmbedding library.

**Comment by jyc** [comment]
Is it just me or does bge-m3 already support running on the GPU with Apple Silicon? The GPU history window in Activity Monitor also shows similar GPU usage whether I run using HuggingFace's sentence_transformers library directly or using this library.

```python
import timeit
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("BAAI/bge-m3")
timeit.timeit(lambda: model.encode(["hi" * 100] * 100), number=10)
# 22.953382124999735
model.encode(["hi"])
# array([[-0.0536464 ,  0.00515981, -0.03156308, ...,  0.01571097,
#         -0.03756364,  0.04886195]], dtype=float32)
```

```python
import timeit
from mlx_embedding_models.embedding import EmbeddingModel
model = EmbeddingModel.from_registry("bge-m3")
timeit.timeit(lambda: model.encode(["hi" * 100] * 100), number=10)
# 25.65148633300032
model.encode(["hi"])
# 100%|██████████████████████████████| 1/1 [00:00<00:00, 11.03it/s, seq_len=16]
# array([[-0.04499852,  0.02141248, -0.02566153, ..., -0.0094963 ,, seq_len=16]
#         -0.01805216,  0.05820245]], dtype=float32)
```

The first output matches the [Colab](https://colab.research.google.com/drive/1bPSenBWpAMewnA2bPXo3-JJOdi4TxXKt#scrollTo=Sa8_XZJu5Wk8) that uses the FlagEmbeddings library:

```
model.encode(["hi"])
# {'dense_vecs': array([[-0.05365,  0.0051 , -0.03156, ...,  0.0157 , -0.03754,  0.04886]],
#        dtype=float16),
#  'lexical_weights': None,
#  'colbert_vecs': None}
```

**Comment by dhruv-anand-aintech** [comment]
here's the notebook for my check: https://colab.research.google.com/drive/1-2bJvoMuYun61QtkBwgmqwwJ0g6TdANI?usp=sharing

**Comment by x4080** [comment]
@jyc is the mlx version faster or use less memory compared to SentenceTransformer using bgem3 ?

Thanks

**Comment by jyc** [comment]
@x4080 No; see the output of timeit that I posted above. I also ran into a pretty severe memory leak which went away when I used the transformers library directly.

**Comment by x4080** [comment]
@jyc  Thanks for quick answer, I tested using onnx and it seems your code is faster (both mlx and sentencetransformers) - so you suggest best using mlx ? with the memory leak in sentencetransformers ?

**Comment by jyc** [comment]
@x4080 Personally I would recommend using the HuggingFace transformers library directly, because ultimately this library and SentenceTransformers are both just wrappers. I don’t think you want to use this library because it does not compute the embeddings correctly, seeing as the outputs differ from the outputs computed using the reference BAAI/bge-m3 code on Colab.

Even if you didn’t care about the outputs matching the reference library, the fact that they don’t match means it’s likely that the properties you’re looking for in a text embedding are not preserved. 

You can use something like this. I extracted this code from my system which has been running for a few weeks now. I haven’t tested the excerpt but any necessary tweaks should be small. 

```python
import torch
from transformers import AutoModel, AutoTokenizer

model = "BAAI/bge-m3"
device = "mps"
tokenizer = AutoTokenizer.from_pretrained(model, device_map=device)
model = AutoModel.from_pretrained(model, device_map=device)
model.eval()

# If you want to save to a local directory, run the following once and then
# replace `model` with `local_path`.
# local_path = "./models/bge-m3"
# model.save_pretrained(local_path)
# tokenizer.save_pretrained(local_path)

encoded = tokenizer(
    input, padding=True, truncation=True, return_tensors="pt"
).to(device)

with torch.no_grad():
    output = model(**encoded)
    embeddings = output[0][:, 0]
    embeddings = torch.nn.functional.normalize(embeddings, p=2, dim=1)
```

**Comment by x4080** [comment]
@jyc Thanks man, very kind of you, I'll test your code

**Comment by jpcompartir** [comment]
@jyc  

Thanks for this - have you benchmarked mps versus mlx for speed*? I may be out of date, but mps was very slow compared to CUDA last time I checked it. Mlx seems more competitive.

I've managed to fix memory leaks for batch embeddings using a custom implementation of MPNet, but I also can't get the results to exactly match the ST/HF results, and the downstream differences are huge (e.g. on clustering tasks or similar)

*clearly if the outputs aren't comparable then speed is basically irrelevant, but we should be able to fix the outputs - or at least get them close enough ( as an aside - different GPUs output slightly different results for embeddings from my experience, whether using ST or HF - it's in the hardware - the question is what's a tolerable difference).

Edit: Ran inference on ~70k documents and MPS outputs were identical to sentence-transformers  with CUDA on a T4 GPU. MLX was ~ 33% faster (2minutes vs 3 minutes for MPS for me) - but outputs are still not identical.  With the above code for torch, MPS was ~ 99% similar to the sentence-transformers outputs with both CUDA and MPS - I suspect the small difference was due to the pooling strategy.
